massa-protocol-worker/src/protocol_worker.rs

```diff
@@ -177,14 +181,27 @@ mod nodeinfo {
             for block_id in block_ids {
                 self.known_blocks.insert(*block_id, (val, instant));
             }
-            while self.known_blocks.len() > max_node_known_blocks_size {
-                // remove oldest item
-                let (&h, _) = self
+            let len = self.known_blocks.len();
+            if len == max_node_known_blocks_size {
+                /* Fast case where there is just one to pop */
+                let block_id = *self
                     .known_blocks
                     .iter()
                     .min_by_key(|(h, (_, t))| (*t, *h))
-                    .unwrap(); // never None because is the collection is empty, while loop isn't executed.
-                self.known_blocks.remove(&h);
+                    .unwrap()
+                    .0;
+                self.known_blocks.remove(&block_id);
+                return;
+            }
+            if len > max_node_known_blocks_size {
+                self.known_blocks
+                    .iter()
+                    .map(|(h, (_, t))| (*h, *t))
+                    .sorted()
```

Comment:
I'd be curious to see how does sorted behave when there is only one item, I'm not sure it's really different of what the fast case does

Reply:
I don't know how the compiler optimize that :rofl: but reading that, the worst case would be O(nlog(n)) and the nice case O(n) where the worst case is the sorted algorithm.

If we go through the nice case with a `while` we could have a kind of (n-m)*n complexity, but visually it looks like a `n!` in the worst case and a O(n) in the nice case.

The question of the robustness surprise me even if I know the pattern now. It's not common to make something robust with the current technics. Normally, or something should never append, so we don't write the _special case_, or it can append, so we choose the better algorithm for the situation.

If we had more than one known block overflow, doesn't that means that we get an error or an undefined behavior earlier? In that case, we should also write a debug comment in addition to a developer note.